### PR TITLE
Parse and check type arguments on JSX opening and self-closing tags

### LINKED ISSUE #6395: Using lookahead to detect generic JSX elements?
_Initially from a [couple](https://github.com/Microsoft/TypeScript/issues/3960#issuecomment-165375803) [different](https://github.com/Microsoft/TypeScript/issues/3960#issuecomment-132192858) comments, but elaborated here._

Currently, in order to use a generic JSX element, you must alias both the type and interface to a non-generic specialization.

``` ts
class Select<T> extends React.Component<SelectProps<T>, any> {}

// Alias
type StringSelect = new () => Select<string>
const StringSelect = <StringSelect> Select

const Form = () => <StringSelect items={['a','b']} />
```

It would be a _lot_ more convenient and consistent to use something like this:

``` ts
const Form = () => <Select<string> items={['a', 'b']} />
```

In order to differentiate the latter, you can use a single token of lookahead in the parser to know if it's a generic JSX element:
- If the next token is a `<`, like in `< Select <`, then consume the rest of the type as a generic type. After consuming that:
  - If the next token is a `>`, then finish the type as a generic type cast.
  - If the next token is an identifier or a `{` (for spread attributes, when they get implemented), then parse the rest as a JSX element.
  - Otherwise, parse the rest as a type cast.
- If the next token is a `>`, like in `< Select >`, then finish the type as a simple type cast.
- If the next token is an identifier or a `{`, then parse the rest as a JSX element.
- Otherwise, consume the rest as a type cast.

I don't know of any reason that wouldn't be feasible to do, considering async arrow functions, which are [already implemented](https://github.com/Microsoft/TypeScript/pull/5231), require potentially a whole call expression of parsing to differentiate between `let f = async (x, y)`, which is a call expression, and `let f = async (x, y) => x`, which is an async arrow function. (I feel sorry for whoever ends up implementing that in V8, which doesn't support any backtracking or much lookahead for memory reasons.)

/cc @RyanCavanaugh

### PR DESCRIPTION
**Pull request by weswigham**
Allows, eg,
```ts
<MyComponent<number> data={12} />
```
in `.tsx` files.

Fixes #6395.

Was there a list of things we were concerned wouldn't parse well? AFAIK, this is actually easy for us to unambiguously parse.

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -17337,6 +17380,17 @@ namespace ts {
...
         }
 
+        function getTypeArgumentArityError(node: Node, signatures: Signature[], typeArguments: NodeArray<TypeNode>) {
+            let min = Number.POSITIVE_INFINITY;
+            let max = Number.NEGATIVE_INFINITY;
+            for (const sig of signatures) {
+                min = Math.min(min, getMinTypeArgumentCount(sig.typeParameters));
+                max = Math.max(max, length(sig.typeParameters));
+            }
+            const paramCount = min < max ? min + "-" + max : min;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: I would find `min === max ? min : min + "-" + max` clearer

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -17337,6 +17380,17 @@ namespace ts {
             }
         }
 
+        function getTypeArgumentArityError(node: Node, signatures: Signature[], typeArguments: NodeArray<TypeNode>) {
+            let min = Number.POSITIVE_INFINITY;
```
Comment (RyanCavanaugh) [inline src/compiler/checker.ts]: Not sure why the original author didn't just use `Infinity` / `-Infinity` ?

**Comment by aaronjensen** [comment]
This is great! Is there a related issue on Babel? What is the process for ensuring they stay up to date?

**Comment by weswigham** [comment]
I don't think we have a formal process right now, but pinging @andy-ms seems like a decent way to check. 😝

**Comment by mvestergaard** [comment]
@weswigham Really great this finally made it into the language!

Question: When tsconfig is set to `"tsx": "preserve"`, is the generic removed?

**Comment by weswigham** [comment]
It aught to be. If not, it's a bug.

**Comment by andrewbranch** [comment]
Is there any sense of in what release this will land?

**Comment by weswigham** [comment]
2.9 which will be released inside two months, given our bimonthly release cadence :smile:

**Comment by marty-wang** [comment]
It seems that one has to explicitly specify the type, as oppose to that the compiler can infer the type from the value. 

What I mean is that in JSX, say I have the component below
```
interface IProps<T> {
    value: T;
    children: (value: T) => React.ReactNode;
}

export class MyComp<T extends {}> extends React.Component<IProps<T>> {
    public render() {
        return this.props.children(this.props.value);
    }
}
```

when I use MyComp, in the code below, specify the type **number**, otherwise the **val** variable will not have type of **number**. 

```
<MyComp<number> value={4}>
    {val => {
       // the type of val is number
        return "";
    }}
</MyComp>
```
However, If I omit specifying the type **number**, the **val** variable will be type of {}.
```
<MyComp value={4}>
    {val => {
       // the type of val is {}
        return "";
    }}
</MyComp>
```

This seems to be a bug, because compare to Generics in regular class, like below., the type of number does not have to be explicitly specified.

```
class A<T> {
    constructor(public val: T) {
    }
}

const a = new A(5)
a.val // val is type of number
```

**Comment by andrewbranch** [comment]
@marty-wang IIUC, I think what you're seeing is #22636

**Comment by jorySsense** [comment]
Question,

What would be best way to pass a generic type to a child component?

```
interface GenericProps<T> {
  isFetching: boolean;
  items: T[];
  removeItem: () => Promise<any>;
  history: History;
}

class ListPage<T> extends React.Component<GenericProps<T>> {
  render() {
    return <LayoutCard {...this.props} />
      <List<T> items={this.props.items}>
        {this.props.children}
      </List>
    </LayoutCard>
  }
};
```

**Comment by weswigham** [comment]
What you have there aught to be correct.

**Comment by jorySsense** [comment]
@weswigham i was getting a 
```
Type '{}' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<ListWithModalConsumer<{}>> & Readonly<{ children?:...'.   Type '{}' is not assignable to type 'Readonly<ListPageProps<{}>>'
```

**Comment by jorySsense** [comment]
Oh i guess i should of mentioned that List accepts a generic type

```
export class List<T> extends React.Component<ListProps<T>> {

  render() {
    const { items, children } = this.props;
    // modal consumer passes showModal hideModal from the contextAPI
    return <ModalConsumer>
      {({ showModal, hideModal }) =>
        <List style={{ width: '100%' }} items={items}
          // list accepts render prop to pass application to list item
              render={(item: T) => children({ item, hideModal, showModal })}>
        </List>
      }
    </ModalConsumer>
  }
}
```

**Comment by weswigham** [comment]
You need to constrained your `T` to at least have the members required by the component you're invoking. In your example, `T` has no constraint, so it is overriding the type implied by your addition of the `items` prop and checking against the `{}` implied by the unconstrained `T`.

**Comment by jameslaneconkling** [comment]
is there a non-jsx equivalent for this?  e.g.

given this usage w/ jsx
```ts
<MyGenericComponent<SomeType> {...props} />
```

what is the equivalent using `createElement`?
```js
createElement(StringInput, props)

// [ts] Type '{}' is not assignable to type 'SomeType'.
```

**Comment by nickmccurdy** [comment]
I believe the first generic parameter of `createElement` is equivalent:
```ts
createElement<SomeType>(MyGenericComponent, props)
```

**Comment by jameslaneconkling** [comment]
The type signature of `createElement` is

```ts
    function createElement<P>(
        type: SFC<P> | ComponentClass<P> | string,
        props?: Attributes & P | null,
        ...children: ReactNode[]): ReactElement<P>;
```

so looks like the first parameter is Props

**Pull request merged**